src/frontend/src/handler/create_source.rs

```diff
@@ -198,31 +198,71 @@ pub(crate) fn bind_all_columns(
     wildcard_idx: Option<usize>,
+    sql_column_strategy: SqlColumnStrategy,
 ) -> Result<Vec<ColumnCatalog>> {
     if let Some(cols_from_source) = cols_from_source {
-        if cols_from_sql.is_empty() {
-            Ok(cols_from_source)
-        } else if let Some(wildcard_idx) = wildcard_idx {
-            if col_defs_from_sql.iter().any(|c| !c.is_generated()) {
-                Err(RwError::from(NotSupported(
-                    "Only generated columns are allowed in user-defined schema from SQL".to_owned(),
-                    "Remove the non-generated columns".to_owned(),
-                )))
-            } else {
-                // Replace `*` with `cols_from_source`
-                let mut cols_from_sql = cols_from_sql;
-                let mut cols_from_source = cols_from_source;
-                let mut cols_from_sql_r = cols_from_sql.split_off(wildcard_idx);
-                cols_from_sql.append(&mut cols_from_source);
-                cols_from_sql.append(&mut cols_from_sql_r);
-                Ok(cols_from_sql)
+        // Need to check `col_defs` to see if a column is generated, as we haven't bind the
+        // `GeneratedColumnDesc` in `ColumnCatalog` yet.
+        let generated_cols_from_sql = cols_from_sql
+            .iter()
+            .filter(|c| {
+                col_defs_from_sql
+                    .iter()
+                    .find(|d| d.name.real_value() == c.name())
+                    .unwrap()
+                    .is_generated()
+            })
+            .cloned()
+            .collect_vec();
+
+        match sql_column_strategy {
+            // Ignore `cols_from_source`, follow `cols_from_sql` without checking.
+            SqlColumnStrategy::Follow => {
+                assert!(
+                    wildcard_idx.is_none(),
+                    "wildcard still exists while strategy is Follows, not correctly purified?"
+                );
+                return Ok(cols_from_sql);
+            }
+
+            // Will merge generated columns from `cols_from_sql` into `cols_from_source`.
+            SqlColumnStrategy::Ignore => {}
+            SqlColumnStrategy::Reject => {
+                // Perform extra check to see if there are non-generated columns in `cols_from_sql`
+                // and reject the request if there are.
+                if generated_cols_from_sql.len() != cols_from_sql.len() {
+                    if wildcard_idx.is_some() {
+                        // (*, normal_column INT)
+                        return Err(RwError::from(NotSupported(
+                            "Only generated columns are allowed in user-defined schema from SQL"
+                                .to_owned(),
+                            "Remove the non-generated columns".to_owned(),
+                        )));
+                    } else {
+                        // (normal_column INT)
+                        // TODO(yuhao): https://github.com/risingwavelabs/risingwave/issues/12209
```

Comment:
Has this TODO been resolved?  #12209  cc. @yuhao-su

Reply:
I think it's more like https://github.com/risingwavelabs/risingwave/issues/12199. There's already `TODO` on `SqlColumnStrategy::Reject`, so I guess we can remove the comments here.